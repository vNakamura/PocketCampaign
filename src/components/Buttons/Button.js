// @flow

import React from 'react';
import styled from 'styled-components';

import type { Theme } from '../../theme';

type Props = {
  active?: boolean,
};
type StyledProps = Props & { theme: Theme };

const StyledButton = styled.button`
  background: none;
  border: none;
  cursor: pointer;
  color: ${(props: StyledProps) => (
      props.active ? props.theme.button.activeColor : props.theme.button.color
    )};
  font-family: ${(props: StyledProps) => props.theme.fonts.display};
  font-size: 1rem;

  &:focus {
    outline: none;
  }
  &:hover,
  &:focus {
    color: ${(props: StyledProps) => props.theme.button.focusColor};
  }
`;
const handleMouseUp = (e: MouseEvent): void => {
  const target: EventTarget = e.currentTarget;
  if (target instanceof HTMLButtonElement) target.blur();
};

const Button = (props: Props) => <StyledButton {...props} onMouseUp={handleMouseUp} />;

Button.defaultProps = {
  active: false,
};

export default Button;
