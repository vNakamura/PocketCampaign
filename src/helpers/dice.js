// @flow

import React from 'react';
import droll from 'droll';
import seedrandom from 'seedrandom';
import replace from 'string-replace-to-array';

import NotationLink from '../components/Chat/NotationLink';

export const notationLimits = {
  minNumDice: 1,
  maxNumDice: 10,
  minNumSides: 2,
  maxNumSides: 100,
  minModifier: -1000,
  maxModifier: 1000,
};
export const diceNotationRegex = /\b([1-9]\d*)?d([1-9]\d*)([+-]\d+)?\b/gi;

type ResultError = {
  ok: false,
  message: string,
};
type ValidateResult = ResultError | { ok: true };
const validateRoll = (notation:string): ValidateResult => {
  const p = droll.parse(notation);
  if (!p) return { ok: false, message: 'Invalid dice notation' };
  if (p.numDice > notationLimits.maxNumDice) return { ok: false, message: 'Invalid dice notation. Too many dice!' };
  if (p.numSides < notationLimits.minNumDice) return { ok: false, message: 'Invalid dice notation. You need at least one die!' };
  if (p.numSides > notationLimits.maxNumSides) return { ok: false, message: 'Invalid dice notation. Too many sides!' };
  if (p.numSides < notationLimits.minNumSides) return { ok: false, message: 'Invalid dice notation. Too few sides!' };
  if (p.modifier > notationLimits.maxModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  if (p.modifier < notationLimits.minModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  return { ok: true };
};

export type ParseResult = ResultError | {
  ok: true,
  parsed: {
    numDice: number,
    numSides: number,
    modifier: number,
    minResult: number,
    maxResult: number,
    avgResult: number,
  }
};
export const parseRoll = (notation: string): ParseResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;
  const p = droll.parse(notation);
  return { ok: true, parsed: p };
};

export type RollResult = ResultError | {
  ok: true,
  rolls: number[],
  modifier: number,
  total: number,
};
export const roll = (notation: string, seed?: string): RollResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;

  if (seed !== undefined) seedrandom(seed, { global: true });
  const result = droll.roll(notation);
  return {
    ok: true,
    rolls: result.rolls,
    modifier: result.modifier,
    total: result.total,
  };
};

export const addLinksToText = (text: string) => {
  let i = 0;
  const result = replace(
    text,
    diceNotationRegex,
    (notation, numDice, numSides) => {
      i += 1;
      const v: ValidateResult = validateRoll(notation);
      if (v.ok && numSides > 1) return <NotationLink key={i} notation={notation} />;
      return notation;
    },
  );
  return result;
};
