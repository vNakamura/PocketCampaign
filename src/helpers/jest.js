import React from 'react';
import { ThemeProvider } from 'styled-components';
/* eslint-disable import/no-extraneous-dependencies */
import renderer from 'react-test-renderer';
import 'jest-styled-components';
/* eslint-enable import/no-extraneous-dependencies */
import theme from '../theme';

export const applyTheme = content => (
  <ThemeProvider theme={theme}>
    {content}
  </ThemeProvider>
);

export const renderWithTheme = content => renderer.create(
  applyTheme(content),
).toJSON();

export default {
  applyTheme,
  renderWithTheme,
};
