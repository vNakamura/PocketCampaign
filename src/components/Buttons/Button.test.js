import React from 'react';

import Button from './Button';
import { renderWithTheme } from '../../helpers/jest';

describe('components: Buttons/Button', () => {
  it('should render correctly', () => {
    const withoutText = renderWithTheme(<Button />);
    expect(withoutText).toMatchSnapshot();
    const withText = renderWithTheme(<Button text={'abcdef'} />);
    expect(withText).toMatchSnapshot();
  });
});
