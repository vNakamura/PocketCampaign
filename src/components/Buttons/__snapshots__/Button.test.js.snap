// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components: Buttons/Button should render correctly 1`] = `
.c0 {
  background: none;
  border: none;
  cursor: pointer;
  color: #817b9e;
  font-family: Dosis;
  font-size: 1rem;
  padding: .2em .4em;
}

.c0:focus {
  outline: none;
}

.c0:hover,
.c0:focus {
  color: #e6e5f1;
}

<button
  className="c0"
  onMouseUp={[Function]}
/>
`;

exports[`components: Buttons/Button should render correctly 2`] = `
.c0 {
  background: none;
  border: none;
  cursor: pointer;
  color: #817b9e;
  font-family: Dosis;
  font-size: 1rem;
  padding: .2em .4em;
}

.c0:focus {
  outline: none;
}

.c0:hover,
.c0:focus {
  color: #e6e5f1;
}

<button
  className="c0"
  onMouseUp={[Function]}
/>
`;
